reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+  }
+}
+
+function IsReadableByteStream(x) {
+  if (!typeIsObject(x)) {
+    return false;
+  }
+
+  if (!Object.prototype.hasOwnProperty.call(x, '_controller')) {
+    return false;
+  }
+
+  return true;
+}
+
+function IsReadableByteStreamByobReader(x) {
+  if (!typeIsObject(x)) {
+    return false;
+  }
+
+  if (!Object.prototype.hasOwnProperty.call(x, '_readIntoRequests')) {
+    return false;
+  }
+
+  return true;
+}
+
+function IsReadableByteStreamController(x) {
+  if (!typeIsObject(x)) {
+    return false;
+  }
+
+  if (!Object.prototype.hasOwnProperty.call(x, '_controlledReadableByteStream')) {
+    return false;
+  }
+
+  return true;
+}
+
+function IsReadableByteStreamLocked(stream) {
+  assert(IsReadableByteStream(stream),
+         'IsReadableByteStreamLocked should only be used on known readable byte streams');
+
+  if (stream._reader === undefined) {
+    return false;
+  }
+
+  return true;
+}
+
+function IsReadableByteStreamReader(x) {
+  if (!typeIsObject(x)) {
+    return false;
+  }
+
+  if (!Object.prototype.hasOwnProperty.call(x, '_readRequests')) {
+    return false;
+  }
+
+  return true;
+}
+
+function PullFromReadableByteStream(stream) {
+  const controller = stream._controller;
+
+  const reader = stream._reader;
+
+  if (reader._readRequests.length > 1) {
+    return;
+  }
+
+  assert(reader._readRequests.length === 1);
+
+  if (controller._totalQueuedBytes === 0) {
+    if (controller._insideUnderlyingByteSource) {
+      controller._callPullOrPullIntoLaterIfNeeded = true;
+      return;
+    }
+
+    CallPull(controller);
+    CallPullOrPullIntoRepeatedlyIfNeeded(controller);
+
+    return;
+  }
+
+  const entry = controller._queue.shift();
+  controller._totalQueuedBytes -= entry.byteLength;
+
+  const view = new Uint8Array(entry.buffer, entry.byteOffset, entry.byteLength);
+
+  const req = reader._readRequests.shift();
+  req.resolve(CreateIterResultObject(view, false));
+
+  if (controller._totalQueuedBytes === 0 && controller._closeRequested) {
+    CloseReadableByteStream(stream);
+  }
+
+  return;
+}
+
+function PullFromReadableByteStreamInto(stream, buffer, byteOffset, byteLength, elementSize) {
+  const controller = stream._controller;
+
+  const pullIntoDescriptor = {
+    buffer,
+    byteOffset,
+    byteLength,
+    bytesFilled: 0,
+    elementSize
+  };
+
+  if (controller._pendingPullIntos.length > 0) {
+    // TODO: Detach buffer.
+    controller._pendingPullIntos.push(pullIntoDescriptor);
+
+    return;
+  }
+
+  if (controller._totalQueuedBytes === 0) {
+    // TODO: Detach buffer.
+    controller._pendingPullIntos.push(pullIntoDescriptor);
+
+    if (controller._insideUnderlyingByteSource) {
+      controller._callPullOrPullIntoLaterIfNeeded = true;
+      return;
+    }
+
+    CallPullInto(controller);
+    CallPullOrPullIntoRepeatedlyIfNeeded(controller);
+
+    return;
+  }
+
+  const ready = FillPullIntoDescriptorFromQueue(controller, pullIntoDescriptor);
```

Comment:
It might be possible to refactor this section to reuse RespondToReadIntoRequestsFromQueue. E.g. push pullIntoDescriptor into _pendingPullIntos and then call it. I am not sure though.